#162948840 implement social authentication with Google, Facebook, and Twitter

Georgeygigz opened the pull request:
**What does this PR do?**
Implement social login with Google, Facebook, and Twitter

Description of Task to be completed?
- Implement social login functionalities for oauth1 and oauth2
- write tests for social auth functionality

**How should this be manually tested?**
- 'git clone https://github.com/andela/ah-django-unchained' 

- `cd ah-django-unchained`

- `git checkout ft-social-authentication-162948840`

- create a virtual env `python3 -m venv virtual_env_name`

- Add the following congigurations to .env file
```
export FACEBOOK_APP_ID="facebook_id"
export FACEBOOK_APP_SECRET="facebook_secret"
export GOOGLE_OAUTH2_KEY="google_key"
export OAUTH2_SECRET="google_secret"
export TWITTER_KEY="twitter_key" 
export TWITTER_SECRET="twitter_secret"
```
- `source .env`

- `pip install -r requirements.txt`

**Start in postman using the following url**
- http://127.0.0.1:8000/api/login/oauth/

**To test different login funcionalities:**

- Facebook
```
{
    "provider":"facebook",
    "access_token":"facebook_access_token"
}
```
- Google:
```
{
    "provider":"google-oauth2",
    "access_token":"google_access_token"
}
```
- Twitter
```
{
    "provider": "twitter",
    "access_token": "access_token",
    "access_token_secret": "access_token_secret"

}
```
**Screenshots**
![image](https://user-images.githubusercontent.com/24627269/51666340-a72d5200-1fce-11e9-8d52-dadf6de6e6d5.png)

![image](https://user-images.githubusercontent.com/24627269/51666352-aeecf680-1fce-11e9-9614-ec2a02a69649.png)


![image](https://user-images.githubusercontent.com/24627269/51666294-911f9180-1fce-11e9-8a1d-1f3b6d69bed5.png)

Georgeygigz committed:
feat(SocialAuthentication):allow user to login using social media

- enable user to login with google, facebook and twitter
- write test for social auth functionality

(Starts #162948840)

ignatiusukwuoma commented on authors/apps/authentication/backends.py:
```diff
@@ -21,32 +24,27 @@ def authenticate(self, request):
+            'exp': int(date.strftime('%s'))
+        }
+        token = jwt.encode(payload, settings.SECRET_KEY, algorithm='HS256')
+        return token.decode()
```
Are you supposed to the token here?

Georgeygigz replied:
Ignatius, Did you meant decode

Georgeygigz committed:
feat(social auth):add social login functionality

- add twitter social media
- handle token generation in backends

[Starts #162948840]

ignatiusukwuoma commented on .gitignore:
```diff
@@ -6,6 +6,8 @@ __pycache__/
 # C extensions
 *.so
 *.vscode
+**/migrations/*
```
It's fine to commit migrations to version control, you don't need these

Georgeygigz replied:
Okay have worked on that

kwameasiago commented:
Why is your Travis failing ?

ignatiusukwuoma commented on authors/apps/authentication/tests/test_social_auth.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class SocialAuthTest(APITestCase):
+    """Test socil authentixation functionality"""
```
typos

Georgeygigz replied:
Okay working on this

Georgeygigz commented:
> Why is your Travis failing?
There is a problem with my environment variables.

Georgeygigz committed:
feat(imports):Remove unused imports

- remove unused imports

[Fixes #162948840]

MaggieChege commented on authors/apps/authentication/views.py:
```diff
@@ -1,24 +1,33 @@
+
+from authors.apps.core.permissions import IsAuthorOrReadOnly
+from authors.apps.authentication.backends import JWTAuthentication
 from . import models
```
This has been imported in line 21. serializers have also been imported twice.

Georgeygigz replied:
Okay fixing it

kwameasiago commented on authors/settings.py:
```diff
@@ -171,7 +209,7 @@
 
 
-SECRET_KEY = os.getenv('SECRET_KEY')
+SECRET_KEY = os.getenv("SECRET_KEY")
```
Hey do not mix ' and " in your code

Georgeygigz replied:
Noted

Georgeygigz committed:
feat(imports):Remove unused imports
- remove double imports
- use single quotes for uniformity

[Fixes #162948840]

ignatiusukwuoma commented on .travis.yml:
```diff
@@ -11,9 +11,15 @@ install:
+  apt:
+    packages:
+      - postgresql-server-dev-9.4
+
```
What does this do?

Georgeygigz replied:
It updates the postgress database to be used in travis

ignatiusukwuoma commented on authors/apps/authentication/migrations/0001_initial.py:
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 2.1.4 on 2019-01-08 14:41
+# Generated by Django 2.1.4 on 2019-01-18 07:38
```
We shouldn't update the migration files

Georgeygigz replied:
Okay noted

ignatiusukwuoma replied:
Can you just remove the change made to this file?

ignatiusukwuoma commented on authors/apps/authentication/serializers.py:
```diff
@@ -195,3 +195,9 @@ def update(self, instance, validated_data):
+class SocialAuthSerializer(serializers.Serializer):
+    """Social auth serializers."""
+    provider = serializers.CharField(max_length=255, required=True)
+    access_token = serializers.CharField(max_length=1024, required=True, trim_whitespace=True)
```
Leave a blank line at the end of all your files

Georgeygigz replied:
Okay will do that

Georgeygigz committed:
feat(add line):add blank line at the end of each file

- add a blank line at the end of each file

[Fixes #162948840]

ignatiusukwuoma commented:
Have you been able to get the user's email? Have you tried making requesting it from the provider using the access-token?

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -151,3 +159,71 @@ def put(self, request, token, **kwargs):
+                 status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+        if user:
+            user.is_active = True
+            username = user.username
```
What happens if username or email already exists. You have to handle it.

Georgeygigz replied:
I have handled this, already

ignatiusukwuoma commented:
I would like to understand what you have in mind on the social login flow. Explain to me, what will happen after user clicks on a button `Login with Facebook` on the frontend

Georgeygigz committed:
feat(Error and Token):Handle errors and add token to variable environments

-Add a descriptive message when user is already authenticated using another account
-Add provider token to virtual environment

[Fixes #162948842]

Georgeygigz committed:
feat(Invalid Token):Return a message when token is invalid

-Custom message for invalid token

[Fixes #162948840]

MaggieChege reviewed (approved):
Look good to me

ignatiusukwuoma commented on .travis.yml:
```diff
@@ -12,11 +12,14 @@ services:
 
 addons:
-  postgresql: "9.6"
+  postgresql: '9.4'
```
I think someone updated it to 9.6. Do you want to downgrade it?

Georgeygigz replied:
No, let me use the 9.6

ignatiusukwuoma commented on .gitignore:
```diff
@@ -26,6 +26,7 @@ var/
 *.egg
 .idea/
 
+
```
Remove this blank line from this file

Georgeygigz replied:
Okay working on it

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -1,25 +1,33 @@
+
 from . import models
+from .renderers import UserJSONRenderer
+from . import serializers
```
Unused import

Georgeygigz replied:
Okay working on it

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -1,25 +1,33 @@
 from rest_framework import status, generics
 from rest_framework.permissions import AllowAny, IsAuthenticated
 from rest_framework.response import Response
 from rest_framework.views import APIView
```
unused import

Georgeygigz replied:
I Am using the import line 246

ignatiusukwuoma commented on authors/apps/authentication/tests/test_social_auth.py:
```diff
@@ -0,0 +1,83 @@
+
+
+class SocialAuthTest(APITestCase):
+    """Test social authentication funcionality functionality"""
```
Remove unnecessary word

Georgeygigz replied:
okay

Georgeygigz committed:
feat(unused imports):Remove unused imports

- Remove imports that are not used

[Fixes #162948840]

ignatiusukwuoma commented on authors/apps/authentication/tests/test_social_auth.py:
```diff
@@ -0,0 +1,83 @@
+        """Test login with facebook."""
+        response = self.client.post(
+            self.social_oauth_url, self.facebook_provider, format='json')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
How about checking that the content of response.data is what was expected for the tests above?

Georgeygigz replied:
Okay will work on that

ignatiusukwuoma commented on authors/apps/authentication/tests/test_social_auth.py:
```diff
@@ -0,0 +1,83 @@
+    def test_rejects_login_missing_provider_name(self):
+        """Test missing provider."""
+        response = self.client.post(self.social_oauth_url,
+                                    data={"access_token": 
```
try to have key-value pairs on the same line except line is too long

Georgeygigz replied:
Okay

Georgeygigz committed:
feat(Check data in response):Check user data returned when user is authenticated

-Remove unnecesary words
-Assert for data in response.data
-Format the code inorder for key-value pairs to be in the same line

[Fixes #162948840]

Georgeygigz committed:
feat(updated travis):upgrade postgress version
-upgrade postgres version

[Fixes #162948840]

Georgeygigz committed:
feat(Facebook config keys):add facebook prefix to facebook keys

-Add facebook prefix to  app_id and secret key

[Fixes #162948845]
